sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -252,6 +259,11 @@ abstract class Optimizer(sessionCatalog: SessionCatalog)
    */
   def extendedOperatorOptimizationRules: Seq[Rule[LogicalPlan]] = Nil
 
+  /**
+   * Override to provide additional rules for early projection and filter pushdown to scans.
+   */
+  def earlyScanPushDownRules: Seq[Rule[LogicalPlan]] = Nil
```

Comment:
can we use the `extendedOperatorOptimizationRules`? It also happens before any rules that depending on stats

Reply:
No.

This needs to be done after filters and projected columns are pushed through operators to colocate them with the scan nodes.